File: fsharp/src/Library/Parser.fs
namespace CallCodingClis

open System.Text.RegularExpressions

type RunnerInfo = {
    Binary: string
    ExtraArgs: string list
    ThinkingFlags: Map<int, string list>
    ProviderFlag: string
    ModelFlag: string
    AgentFlag: string
}

type ParsedArgs = {
    Runner: string option
    Thinking: int option
    Provider: string option
    Model: string option
    Alias: string option
    Prompt: string
}

type AliasDef = {
    Runner: string option
    Thinking: int option
    Provider: string option
    Model: string option
    Agent: string option
}

type CccConfig = {
    DefaultRunner: string
    DefaultProvider: string
    DefaultModel: string
    DefaultThinking: int option
    Aliases: Map<string, AliasDef>
    Abbreviations: Map<string, string>
}

module Parser =

    let runnerRegistry : Map<string, RunnerInfo> =
        let openCodeInfo = {
            Binary = "opencode"
            ExtraArgs = ["run"]
            ThinkingFlags = Map.empty
            ProviderFlag = ""
            ModelFlag = ""
            AgentFlag = "--agent"
        }
        let claudeInfo = {
            Binary = "claude"
            ExtraArgs = []
            ThinkingFlags = Map.ofList [
                0, ["--thinking"; "disabled"]
                1, ["--thinking"; "enabled"; "--effort"; "low"]
                2, ["--thinking"; "enabled"; "--effort"; "medium"]
                3, ["--thinking"; "enabled"; "--effort"; "high"]
                4, ["--thinking"; "enabled"; "--effort"; "max"]
            ]
            ProviderFlag = ""
            ModelFlag = "--model"
            AgentFlag = "--agent"
        }
        let kimiInfo = {
            Binary = "kimi"
            ExtraArgs = []
            ThinkingFlags = Map.ofList [
                0, ["--no-thinking"]
                1, ["--thinking"]
                2, ["--thinking"]
                3, ["--thinking"]
                4, ["--thinking"]
            ]
            ProviderFlag = ""
            ModelFlag = "--model"
            AgentFlag = "--agent"
        }
        let codexInfo = {
            Binary = "codex"
            ExtraArgs = []
            ThinkingFlags = Map.empty
            ProviderFlag = ""
            ModelFlag = "--model"
            AgentFlag = ""
        }
        let roocodeInfo = {
            Binary = "roocode"
            ExtraArgs = []
            ThinkingFlags = Map.empty
            ProviderFlag = ""
            ModelFlag = ""
            AgentFlag = ""
        }
        let crushInfo = {
            Binary = "crush"
            ExtraArgs = []
            ThinkingFlags = Map.empty
            ProviderFlag = ""
            ModelFlag = ""
            AgentFlag = ""
        }
        Map.ofList [
            "opencode", openCodeInfo
            "claude", claudeInfo
            "kimi", kimiInfo
            "codex", codexInfo
            "roocode", roocodeInfo
            "crush", crushInfo
            "oc", openCodeInfo
            "cc", claudeInfo
            "c", codexInfo
            "cx", codexInfo
            "k", kimiInfo
            "rc", roocodeInfo
            "cr", crushInfo
        ]

    let private runnerSelectorRe = Regex(@"^(?:oc|cc|cx|c|k|rc|cr|codex|claude|opencode|kimi|roocode|crush|pi)$", RegexOptions.IgnoreCase)
    let private thinkingRe = Regex(@"^\+([0-4])$")
    let private providerModelRe = Regex(@"^:([a-zA-Z0-9_-]+):([a-zA-Z0-9._-]+)$")
    let private modelRe = Regex(@"^:([a-zA-Z0-9._-]+)$")
    let private aliasRe = Regex(@"^@([a-zA-Z0-9_-]+)$")

    let parseArgs (argv: string array) : ParsedArgs =
        let mutable runner = None
        let mutable thinking = None
        let mutable provider = None
        let mutable model = None
        let mutable aliasName = None
        let positional = ResizeArray<string>()

        for token in argv do
            if runnerSelectorRe.IsMatch(token) && runner.IsNone && positional.Count = 0 then
                runner <- Some(token.ToLower())
            elif thinkingRe.IsMatch(token) && positional.Count = 0 then
                let m = thinkingRe.Match(token)
                thinking <- Some(int m.Groups.[1].Value)
            elif providerModelRe.IsMatch(token) && positional.Count = 0 then
                let m = providerModelRe.Match(token)
                provider <- Some(m.Groups.[1].Value)
                model <- Some(m.Groups.[2].Value)
            elif modelRe.IsMatch(token) && positional.Count = 0 then
                let m = modelRe.Match(token)
                model <- Some(m.Groups.[1].Value)
            elif aliasRe.IsMatch(token) && aliasName.IsNone && positional.Count = 0 then
                let m = aliasRe.Match(token)
                aliasName <- Some(m.Groups.[1].Value)
            else
                positional.Add(token)

        { Runner = runner
          Thinking = thinking
          Provider = provider
          Model = model
          Alias = aliasName
          Prompt = System.String.Join(" ", positional) }

    let private resolveRunnerName (name: string option) (config: CccConfig) : string =
        match name with
        | None -> config.DefaultRunner
        | Some n ->
            match config.Abbreviations.TryFind(n) with
            | Some abbrev -> abbrev
            | None -> n

    let resolveCommand (parsed: ParsedArgs) (config: CccConfig option) : string list * Map<string, string> * string list =
        let config = defaultArg config {
            DefaultRunner = "oc"
            DefaultProvider = ""
            DefaultModel = ""
            DefaultThinking = None
            Aliases = Map.empty
            Abbreviations = Map.empty
        }

        let runnerName = resolveRunnerName parsed.Runner config
        let mutable warnings = []

        let openCodeInfo =
            match runnerRegistry.TryFind("opencode") with
            | Some info -> info
            | None -> {
                Binary = "opencode"
                ExtraArgs = ["run"]
                ThinkingFlags = Map.empty
                ProviderFlag = ""
                ModelFlag = ""
                AgentFlag = "--agent"
            }

        let fallbackInfo =
            match runnerRegistry.TryFind(config.DefaultRunner) with
            | Some info -> info
            | None -> openCodeInfo

        let info =
            match runnerRegistry.TryFind(runnerName) with
            | Some i -> i
            | None -> fallbackInfo

        let mutable effectiveRunnerName = runnerName

        let aliasDef =
            match parsed.Alias with
            | Some a ->
                match config.Aliases.TryFind(a) with
                | Some ad -> Some ad
                | None -> None
            | None -> None

        let info =
            match aliasDef with
            | Some ad ->
                match ad.Runner with
                | Some r when parsed.Runner.IsNone ->
                    let ern = resolveRunnerName (Some r) config
                    effectiveRunnerName <- ern
                    match runnerRegistry.TryFind(ern) with Some i -> i | None -> info
                | _ -> info
            | None -> info

        let argv = ResizeArray<string>()
        argv.Add(info.Binary)
        for a in info.ExtraArgs do argv.Add(a)

        let effectiveThinking =
            match parsed.Thinking with
            | Some t -> Some t
            | None ->
                match aliasDef with
                | Some ad -> ad.Thinking
                | None -> config.DefaultThinking

        match effectiveThinking with
        | Some t ->
            match info.ThinkingFlags.TryFind(t) with
            | Some flags -> for f in flags do argv.Add(f)
            | None -> ()
        | None -> ()

        let effectiveProvider =
            match parsed.Provider with
            | Some p -> Some p
            | None ->
                match aliasDef with
                | Some ad -> ad.Provider
                | None ->
                    if System.String.IsNullOrEmpty(config.DefaultProvider) then None
                    else Some config.DefaultProvider

        let effectiveModel =
            match parsed.Model with
            | Some m -> Some m
            | None ->
                match aliasDef with
                | Some ad -> ad.Model
                | None ->
                    if System.String.IsNullOrEmpty(config.DefaultModel) then None
                    else Some config.DefaultModel

        match effectiveModel with
        | Some m when not (System.String.IsNullOrEmpty(info.ModelFlag)) ->
            argv.Add(info.ModelFlag)
            argv.Add(m)
        | _ -> ()

        let effectiveAgent =
            match parsed.Alias with
            | Some alias ->
                match aliasDef with
                | Some ad when ad.Agent.IsSome && not (System.String.IsNullOrEmpty ad.Agent.Value) ->
                    ad.Agent
                | Some _ ->
                    None
                | None -> Some alias
            | None -> None

        match effectiveAgent with
        | Some agent when not (System.String.IsNullOrEmpty agent) ->
            if not (System.String.IsNullOrEmpty info.AgentFlag) then
                argv.Add(info.AgentFlag)
                argv.Add(agent)
            else
                warnings <- warnings @ [sprintf "warning: runner \"%s\" does not support agents; ignoring @%s" effectiveRunnerName agent]
        | _ -> ()

        let envOverrides = ResizeArray<string * string>()
        match effectiveProvider with
        | Some p -> envOverrides.Add(("CCC_PROVIDER", p))
        | None -> ()

        let prompt = parsed.Prompt.Trim()
        if System.String.IsNullOrEmpty(prompt) then
            failwith "prompt must not be empty"

        argv.Add(prompt)
        List.ofSeq argv, Map.ofSeq envOverrides, warnings

File: haskell/src/CallCodingClis/Parser.hs
module CallCodingClis.Parser
  ( RunnerInfo(..)
  , ParsedArgs(..)
  , AliasDef(..)
  , CccConfig(..)
  , runnerRegistry
  , defaultConfig
  , parseArgs
  , resolveCommand
  ) where

import Control.Applicative ((<|>))
import Data.Char (isSpace, toLower)
import Data.List (dropWhileEnd)
import Data.Map.Strict (Map)
import qualified Data.Map.Strict as Map
import Data.Maybe (fromMaybe)

data RunnerInfo = RunnerInfo
  { riBinary        :: String
  , riExtraArgs     :: [String]
  , riThinkingFlags :: Map Int [String]
  , riProviderFlag  :: String
  , riModelFlag     :: String
  , riAgentFlag     :: String
  } deriving (Eq, Show)

data ParsedArgs = ParsedArgs
  { paRunner   :: Maybe String
  , paThinking :: Maybe Int
  , paProvider :: Maybe String
  , paModel    :: Maybe String
  , paAlias    :: Maybe String
  , paPrompt   :: String
  } deriving (Eq, Show)

data AliasDef = AliasDef
  { adRunner   :: Maybe String
  , adThinking :: Maybe Int
  , adProvider :: Maybe String
  , adModel    :: Maybe String
  , adAgent    :: Maybe String
  } deriving (Eq, Show)

data CccConfig = CccConfig
  { ccDefaultRunner   :: String
  , ccDefaultProvider :: String
  , ccDefaultModel    :: String
  , ccDefaultThinking :: Maybe Int
  , ccAliases         :: Map String AliasDef
  , ccAbbreviations   :: Map String String
  } deriving (Eq, Show)

runnerRegistry :: Map String RunnerInfo
runnerRegistry = Map.fromList $ baseRunners ++ abbrevs
  where
    opencode = RunnerInfo "opencode" ["run"] Map.empty "" "" "--agent"
    claude = RunnerInfo "claude" [] (Map.fromList
      [ (0, ["--no-thinking"])
      , (1, ["--thinking", "low"])
      , (2, ["--thinking", "medium"])
      , (3, ["--thinking", "high"])
      , (4, ["--thinking", "max"])
      ]) "" "--model" "--agent"
    kimi = RunnerInfo "kimi" [] (Map.fromList
      [ (0, ["--no-think"])
      , (1, ["--think", "low"])
      , (2, ["--think", "medium"])
      , (3, ["--think", "high"])
      , (4, ["--think", "max"])
      ]) "" "--model" "--agent"
    codex = RunnerInfo "codex" [] Map.empty "" "--model" ""
    roocode = RunnerInfo "roocode" [] Map.empty "" "--model" ""
    crush = RunnerInfo "crush" [] Map.empty "" "" ""

    baseRunners =
      [ ("opencode", opencode)
      , ("claude", claude)
      , ("kimi", kimi)
      , ("codex", codex)
      , ("roocode", roocode)
      , ("crush", crush)
      ]

    abbrevs =
      [ ("oc", opencode)
      , ("cc", claude)
      , ("c", codex)
      , ("cx", codex)
      , ("k", kimi)
      , ("rc", roocode)
      , ("cr", crush)
      ]

defaultConfig :: CccConfig
defaultConfig = CccConfig "oc" "" "" Nothing Map.empty Map.empty

runnerSelectorSet :: [String]
runnerSelectorSet =
  ["oc","cc","c","cx","k","rc","cr","codex","claude","opencode","kimi","roocode","crush","pi"]

isRunnerSelector :: String -> Bool
isRunnerSelector t = map toLower t `elem` runnerSelectorSet

matchThinking :: String -> Maybe Int
matchThinking ('+':[d])
  | d >= '0' && d <= '4' = Just (fromEnum d - fromEnum '0')
matchThinking _ = Nothing

isAsciiAlphaNum :: Char -> Bool
isAsciiAlphaNum c =
  (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9')

matchProviderModel :: String -> Maybe (String, String)
matchProviderModel (':':rest) = case break (== ':') rest of
  (provider, ':':model) ->
    let okProvider = not (null provider) && all (\c -> isAsciiAlphaNum c || c == '-' || c == '_') provider
        okModel   = not (null model)    && all (\c -> isAsciiAlphaNum c || c == '.' || c == '-' || c == '_') model
    in if okProvider && okModel then Just (provider, model) else Nothing
  _ -> Nothing
matchProviderModel _ = Nothing

matchModel :: String -> Maybe String
matchModel (':':model) =
  let ok = not (null model) && all (\c -> isAsciiAlphaNum c || c == '.' || c == '-' || c == '_') model
  in if ok then Just model else Nothing
matchModel _ = Nothing

matchAlias :: String -> Maybe String
matchAlias ('@':alias) =
  let ok = not (null alias) && all (\c -> isAsciiAlphaNum c || c == '-' || c == '_') alias
  in if ok then Just alias else Nothing
matchAlias _ = Nothing

parseArgs :: [String] -> ParsedArgs
parseArgs = go (ParsedArgs Nothing Nothing Nothing Nothing Nothing "") []
  where
    go parsed positional [] = parsed { paPrompt = unwords positional }
    go parsed positional (token:tokens)
      | isRunnerSelector token
      , Nothing <- paRunner parsed
      , null positional =
          go parsed { paRunner = Just (map toLower token) } positional tokens
      | Just n <- matchThinking token
      , null positional =
          go parsed { paThinking = Just n } positional tokens
      | Just (prov, mod) <- matchProviderModel token
      , null positional =
          go parsed { paProvider = Just prov, paModel = Just mod } positional tokens
      | Just mod <- matchModel token
      , null positional =
          go parsed { paModel = Just mod } positional tokens
      | Just alias <- matchAlias token
      , Nothing <- paAlias parsed
      , null positional =
          go parsed { paAlias = Just alias } positional tokens
      | otherwise =
          go parsed (positional ++ [token]) tokens

strip :: String -> String
strip = dropWhileEnd isSpace . dropWhile isSpace

resolveRunnerName :: Maybe String -> CccConfig -> String
resolveRunnerName Nothing config    = ccDefaultRunner config
resolveRunnerName (Just name) config =
  case Map.lookup name (ccAbbreviations config) of
    Just abbrev -> abbrev
    Nothing     -> name

nonEmpty :: String -> Maybe String
nonEmpty s
  | null s    = Nothing
  | otherwise = Just s

resolveCommand :: ParsedArgs -> Maybe CccConfig -> Either String ([String], Map String String, [String])
resolveCommand parsed mConfig =
  let
    config = fromMaybe defaultConfig mConfig
    warnings = []

    runnerName = resolveRunnerName (paRunner parsed) config

    opencodeInfo = runnerRegistry Map.! "opencode"
    info = fromMaybe opencodeInfo $
      Map.lookup runnerName runnerRegistry <|>
      Map.lookup (ccDefaultRunner config) runnerRegistry

    aliasDef = paAlias parsed >>= \a -> Map.lookup a (ccAliases config)

    effectiveInfo = case (aliasDef >>= adRunner, paRunner parsed) of
      (Just ar, Nothing) ->
        let name = resolveRunnerName (Just ar) config
        in fromMaybe info (Map.lookup name runnerRegistry)
      _ -> info

    baseArgv = riBinary effectiveInfo : riExtraArgs effectiveInfo

    effectiveThinking =
      paThinking parsed <|>
      (aliasDef >>= adThinking) <|>
      ccDefaultThinking config

    argvWithThinking = case effectiveThinking of
      Just n  -> case Map.lookup n (riThinkingFlags effectiveInfo) of
        Just flags -> baseArgv ++ flags
        Nothing    -> baseArgv
      Nothing -> baseArgv

    effectiveProvider =
      paProvider parsed <|>
      (aliasDef >>= adProvider) <|>
      nonEmpty (ccDefaultProvider config)

    effectiveModel =
      paModel parsed <|>
      (aliasDef >>= adModel) <|>
      nonEmpty (ccDefaultModel config)

    effectiveAgent = case paAlias parsed of
      Nothing    -> Nothing
      Just alias ->
        case aliasDef >>= adAgent of
          Just agent -> Just agent
          Nothing    -> if Map.member alias (ccAliases config) then Nothing else Just alias

    argvWithModel = case effectiveModel of
      Just m | not (null (riModelFlag effectiveInfo)) ->
        argvWithThinking ++ [riModelFlag effectiveInfo, m]
      _ -> argvWithThinking

    (argvWithAgent, warnings') = case effectiveAgent of
      Just agent | not (null (riAgentFlag effectiveInfo)) ->
        (argvWithModel ++ [riAgentFlag effectiveInfo, agent], warnings)
      Just agent ->
        let warning = "warning: runner \"" ++ riBinary effectiveInfo ++ "\" does not support agents; ignoring @" ++ agent
        in (argvWithModel, warnings ++ [warning])
      Nothing -> (argvWithModel, warnings)

    envOverrides = case effectiveProvider of
      Just p  -> Map.singleton "CCC_PROVIDER" p
      Nothing -> Map.empty

    prompt = strip (paPrompt parsed)
  in
    if null prompt
      then Left "prompt must not be empty"
      else Right (argvWithAgent ++ [prompt], envOverrides, warnings')
